File: 994. 腐烂的橘子.py
import collections


class Solution(object):
    def orangesRotting(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        m = len(grid)
        n = len(grid[0])
        q = collections.deque()
        fresh_count = 0
        minites = 0
        directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]

        for i in range(m):
            for j in range(n):
                if grid[i][j] == 2:
                    q.append((i, j))
                elif grid[i][j] == 1:
                    fresh_count += 1

        while q and fresh_count != 0:
            minites += 1
            for _ in range(len(q)):
                curr_i, curr_j = q.popleft()
                for x, y in directions:
                    temp_i, temp_j = curr_i + x, curr_j + y
                    if 0 <= temp_i < m and 0 <= temp_j < n and grid[temp_i][temp_j] == 1:
                        grid[temp_i][temp_j] = 2
                        fresh_count -= 1
                        q.append((temp_i, temp_j))
        return minites if fresh_count == 0 else -1

File: 994. 腐烂的橘子.go
package main

func orangesRotting(grid [][]int) int {
	minutes := 0
	freshCount := 0
	m, n := len(grid), len(grid[0])
	q := make([][]int, 0)

	for i := range m {
		for j := range n {
			if grid[i][j] == 1 {
				freshCount++
			}
			if grid[i][j] == 2 {
				q = append(q, []int{i, j})
			}
		}
	}

	var infect func(i, j int)
	infect = func(i, j int) {
		if i < 0 || i >= m || j < 0 || j >= n {
			return
		}
		if grid[i][j] == 1 {
			grid[i][j] = 2
			freshCount--
			q = append(q, []int{i, j})
		}
	}
	var spread func(i, j int)
	spread = func(i, j int) {
		infect(i+1, j)
		infect(i-1, j)
		infect(i, j+1)
		infect(i, j-1)
	}

	for freshCount != 0 && len(q) != 0 {
		minutes += 1
		length := len(q)
		for idx := range length {
			badLoc := q[idx]
			spread(badLoc[0], badLoc[1])
		}
		q = q[length:]
	}

	if freshCount == 0 {
		return minutes
	}
	return -1
}
